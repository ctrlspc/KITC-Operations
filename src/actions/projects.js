import database from '../firebase/firebase';

export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const startSetProjects = () => {
  return (dispatch) => {
    return database.ref(`projects`)
      .once('value')
      .then((snapshot) => {
        const projects = [];

        snapshot.forEach(childsnapshot => {
          projects.push({
            id:childsnapshot.key,
            ...childsnapshot.val()
          });
        });        
        dispatch(setProjects(projects));
      });
  };
};

export const addProject = (project) => ({
  type: 'ADD_PROJECT',
  project
});

export const startAddProject = ({title, description, projectManager, projectType}) => {
 return (dispatch) => {
    const project = {title, description, projectManager, projectType};

    return database.ref(`projects`)
      .push(project)
      .then((ref) => {
        dispatch(addProject({
          id:ref.key,
          ...project
        }));

        return Promise.resolve(ref.key);
      });
  };
};

export const updateProject = (id, updates) => ({
  type: 'UPDATE_PROJECT',
  id, 
  updates
});

export const startUpdateProject = (id, updates) => {
  return (dispatch) => {
    return database.ref(`projects/${id}`)
      .update(updates)
      .then((ref) => {
        dispatch(updateProject(id, updates));
      });
    };
};
