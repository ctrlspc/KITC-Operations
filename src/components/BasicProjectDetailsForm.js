import React from 'react';

export default class BasicProjectDetailsForm extends React.Component {
 
  constructor(props) {
    super(props);
    const project = this.props.project;
    this.state = {
      title: project ? project.title : '',
      description: project ? project.description : '',
      projectManager: project ? project.projectManager  : this.props.projectManagers[0] ,
      projectType: project ? project.projectType : 'ext',
      error:''
    };
  };

  onTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({title}));
  };

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description}));
  };

  onManagerChange = (e) => {
    const projectManager = this.props.projectManagers.find((projectManager) => {
      return projectManager.id === e.target.value
    });
    this.setState(() => ({projectManager}));
  };

  onTypeChange = (e) => {
    const projectType = e.target.value;
    this.setState(() => ({projectType}));
  };

  onSubmit = (e) => {
    e.preventDefault();

    if(!this.state.title || !this.state.description || !this.state.projectManager) {
      this.setState(() => ({error:'Please provide a title and project description'}));
    } else {
      this.setState(() => ({error:''}));
      this.props.onSubmit({
        title: this.state.title,
        description: this.state.description,
        projectManager: this.state.projectManager ,
        projectType: this.state.projectType
      });
    }
  };

  render () {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p> }
        <input 
          autoFocus
          className="text-input"
          type="text" 
          placeholder="Project Name"
          value={this.state.title}
          onChange={this.onTitleChange}
        />
        <textarea 
          className="textarea"
          placeholder="Project Description"
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <div className="select-group">
          <div className="select-group__item">
            <h3 className="select-group__title">Project Manager</h3>
            <select 
              className="select-group__select"
              id="project-manager"
              onChange={this.onManagerChange}
              value={!!this.state.projectManager ? this.state.projectManager.id : ''}
            >
              {this.props.projectManagers && this.props.projectManagers.map((item) => {
                return <option key={item.id} value={item.id}>{item.name}</option> 
              })}
            </select>
          </div>
          <div className="select-group__item">
            <h3 className="select-group__title">Project Type</h3>
            <select 
              className="select-group__select"
              id="project-type" 
              onChange={this.onTypeChange}
              value={this.state.projectType}
            >
              <option value="int">Internal</option>
              <option value="ext">External</option>
            </select>
          </div>
        </div>
        <div>
          <button className="button">Save Project</button>
        </div>
        
      </form>
    );
  };
};
