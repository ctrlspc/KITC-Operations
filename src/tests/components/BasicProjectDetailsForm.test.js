import React from 'react';
import { shallow } from 'enzyme';
import BasicProjectDetailsForm from '../../components/BasicProjectDetailsForm';
import projects from '../fixtures/projects';
import profiles from '../fixtures/profiles';


test('should render form correctly with no project data', () => {
  const wrapper = shallow(<BasicProjectDetailsForm projectManagers={profiles}/>)
  expect(wrapper).toMatchSnapshot();
});

test('should render form correctly with project data passed in', () => {
  const wrapper = shallow(<BasicProjectDetailsForm project={projects[0]} projectManagers={profiles}/>)
  expect(wrapper).toMatchSnapshot();
});

test('should set the project title', () => {
  const wrapper = shallow(<BasicProjectDetailsForm projectManagers={profiles} />);
  const value = 'New Title';

  wrapper.find('input').at(0).simulate('change', {
    target: {value}
  });

  expect(wrapper.state('title')).toBe(value);
});

test('should set the project description', () => {
  const wrapper = shallow(<BasicProjectDetailsForm projectManagers={profiles}/>);
  const value = 'New Description';

  wrapper.find('textarea').at(0).simulate('change', {
    target: {value}
  });

  expect(wrapper.state('description')).toBe(value);
});

test('should set the project manager', () => {
  const wrapper = shallow(<BasicProjectDetailsForm projectManagers={profiles}/>);
  const value = profiles[0].uid;

  wrapper.find('#project-manager').at(0).simulate('change', {
    target: {value}
  });

  expect(wrapper.state('projectManager')).toBe(profiles[0].uid);
});

test('should set the project type', () => {
  const wrapper = shallow(<BasicProjectDetailsForm projectManagers={profiles}/>);
  const value = 'int';

  wrapper.find('#project-type').at(0).simulate('change', {
    target: {value}
  });

  expect(wrapper.state('projectType')).toBe(value);
});

test('should call onSubmit for a valid submission', () => {
  const onSubmitSpy = jest.fn();
  const preventDefaultSpy = jest.fn();
  const wrapper = shallow(
    <BasicProjectDetailsForm 
      project={projects[0]} 
      onSubmit={onSubmitSpy}
      projectManagers={profiles}
    />);
  wrapper.find('form').simulate('submit',{preventDefault:preventDefaultSpy});
  expect(preventDefaultSpy).toHaveBeenCalled();
  expect(wrapper.state('error')).toBe('');
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    title: projects[0].title,
    description: projects[0].description,
    projectManager: profiles[0],
    projectType: projects[0].projectType
  });
});

test('should set an error for an invalid submission', () => {
  const wrapper = shallow(<BasicProjectDetailsForm projectManagers={profiles}/>);
  wrapper.find('form').simulate('submit',{preventDefault:() => {}});
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});