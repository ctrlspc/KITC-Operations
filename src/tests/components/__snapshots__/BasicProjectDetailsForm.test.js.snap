// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render form correctly with no project data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Project Name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Project Description"
    value=""
  />
  <div
    className="select-group"
  >
    <div
      className="select-group__item"
    >
      <h3
        className="select-group__title"
      >
        Project Manager
      </h3>
      <select
        className="select-group__select"
        id="project-manager"
        onChange={[Function]}
        value="1"
      >
        <option
          key="1"
          value="1"
        >
          Mr Test
        </option>
        <option
          key="2"
          value="2"
        >
          Mr Testy
        </option>
        <option
          key="3"
          value="3"
        >
          Mr Grumpy
        </option>
      </select>
    </div>
    <div
      className="select-group__item"
    >
      <h3
        className="select-group__title"
      >
        Project Type
      </h3>
      <select
        className="select-group__select"
        id="project-type"
        onChange={[Function]}
        value="ext"
      >
        <option
          value="int"
        >
          Internal
        </option>
        <option
          value="ext"
        >
          External
        </option>
      </select>
    </div>
  </div>
  <div>
    <button
      className="button"
    >
      Save Project
    </button>
  </div>
</form>
`;

exports[`should render form correctly with project data passed in 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Project Name"
    type="text"
    value="first test project"
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Project Description"
    value="testing some great stuff"
  />
  <div
    className="select-group"
  >
    <div
      className="select-group__item"
    >
      <h3
        className="select-group__title"
      >
        Project Manager
      </h3>
      <select
        className="select-group__select"
        id="project-manager"
        onChange={[Function]}
        value="1"
      >
        <option
          key="1"
          value="1"
        >
          Mr Test
        </option>
        <option
          key="2"
          value="2"
        >
          Mr Testy
        </option>
        <option
          key="3"
          value="3"
        >
          Mr Grumpy
        </option>
      </select>
    </div>
    <div
      className="select-group__item"
    >
      <h3
        className="select-group__title"
      >
        Project Type
      </h3>
      <select
        className="select-group__select"
        id="project-type"
        onChange={[Function]}
        value="int"
      >
        <option
          value="int"
        >
          Internal
        </option>
        <option
          value="ext"
        >
          External
        </option>
      </select>
    </div>
  </div>
  <div>
    <button
      className="button"
    >
      Save Project
    </button>
  </div>
</form>
`;

exports[`should set an error for an invalid submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Please provide a title and project description
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Project Name"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Project Description"
    value=""
  />
  <div
    className="select-group"
  >
    <div
      className="select-group__item"
    >
      <h3
        className="select-group__title"
      >
        Project Manager
      </h3>
      <select
        className="select-group__select"
        id="project-manager"
        onChange={[Function]}
        value="1"
      >
        <option
          key="1"
          value="1"
        >
          Mr Test
        </option>
        <option
          key="2"
          value="2"
        >
          Mr Testy
        </option>
        <option
          key="3"
          value="3"
        >
          Mr Grumpy
        </option>
      </select>
    </div>
    <div
      className="select-group__item"
    >
      <h3
        className="select-group__title"
      >
        Project Type
      </h3>
      <select
        className="select-group__select"
        id="project-type"
        onChange={[Function]}
        value="ext"
      >
        <option
          value="int"
        >
          Internal
        </option>
        <option
          value="ext"
        >
          External
        </option>
      </select>
    </div>
  </div>
  <div>
    <button
      className="button"
    >
      Save Project
    </button>
  </div>
</form>
`;
